app/controllers/application_controller.rb

```diff
@@ -57,7 +57,32 @@ def analytics_user
   end
 
   def create_user_event(event_type, user = current_user)
-    Event.create(user_id: user.id, event_type: event_type)
+    return unless user&.id
+    device = create_or_update_device(user)
+    Event.create(user_id: user.id,
+                 device_id: device.id,
+                 ip: request.remote_ip,
+                 event_type: event_type)
+  end
+
+  def create_or_update_device(user)
+    device = DeviceTracking::HasDevice.call(user, cookies[:device])
+    if device
+      DeviceTracking::UpdateDevice.call(device, request)
+    else
+      device = create_device(user)
+    end
+    device
+  end
+
+  def create_device(user)
+    cookie_uuid = cookies[:device]
+    device = DeviceTracking::CreateDevice.call(user, request, cookie_uuid)
+    device_uuid = device.cookie_uuid
+    unless device_uuid == cookie_uuid
+      cookies[:device] = { value: device_uuid, expires: 1.year.from_now }
```

Comment:
If you don't specify a expiration, the cookie should be permanent, right? Isn't that what we want instead of it expiring and becoming a new device one day a year from now?

Reply:
If no expiration is supplied the cookie will be deleted when the browser is closed.  To create a permanent cookie you set the expiration date far into the future.  cookie.permanent also does this (using 20 years from now).  I'll just add that.